src/js_parser.zig

```diff
@@ -6296,6 +6087,105 @@ fn NewParser_(
             }) catch unreachable;
         }
 
+        pub fn generateReactRefreshImport(p: *P, parts: *ListManaged(js_ast.Part)) !void {
+            switch (p.options.features.hot_module_reloading) {
+                inline else => |hmr| try p.generateReactRefreshImportHmr(parts, hmr),
+            }
+        }
+
+        fn generateReactRefreshImportHmr(p: *P, parts: *ListManaged(js_ast.Part), comptime hot_module_reloading: bool) !void {
+            // If `hot_module_reloading`, we are going to generate a require call:
+            //
+            //     const { $RefreshSig$, $RefreshReg$ } = require("react-refresh/runtime")`
+            //
+            // Otherwise we are going to settle on an import statement. Using
+            // require is fine in HMR bundling because `react-refresh` itself is
+            // already a CommonJS module, and it will actually be more efficient
+            // at runtime this way.
+            const allocator = p.allocator;
+            const import_record_index = p.addImportRecordByRange(.stmt, logger.Range.None, "react-refresh/runtime");
+
+            const Item = if (hot_module_reloading) B.Object.Property else js_ast.ClauseItem;
+
```

Comment:
`const num_items = 1 + @intFromBool(p.react_refresh.register_used) + @intFromBool(p.react_refresh.signature_used);` and use it for `items.initCapacity` and `declared_symbols.ensureTotalCapacity`?